reconstructSPI/iterative_refinement/expectation_maximization.py

```diff
@@ -274,7 +274,31 @@ def grid_SO3_uniform(n_rotations):
             Array describing rotations.
             Shape (n_rotations, 3, 3)
         """
-        rots = np.ones((n_rotations, 3, 3))
+        rots = np.zeros((n_rotations, 3, 3))
+        for i in range(n_rotations):
+            phi, theta, psi = np.random.random(3) * 2 * np.pi
```

Comment:
This does not generate uniform rotations. They might look uniform in the picture, but they bunch at the poll. If you plotted a histogram on the sphere (like sky scrapers on the face of the globe) they would be higher on the polls. 

Uniform is defined not in the parameters of the encoding of SO(3). I like to think of uniform on SO(3) as being defined as the distance between rotations tending to be equal. This raises the question of what distance is in SO(3).  See the equation in "1. Introduction: Single Rotation Averaging" in  Hartley, R., Trumpf, J., Dai, Y., & Li, H. (2013). Rotation Averaging. International Journal of Computer Vision, 103(3), 267–305. http://doi.org/10.1007/s11263-012-0601-0

You can use random_uniform in geomstats
https://geomstats.github.io/_modules/geomstats/geometry/special_orthogonal.html

I did something here that is uniform, but you should be able to just use what's in geomstats, not implement anything new yourselves. The less you implement, the less you have to test!
https://github.com/geoffwoollard/compSPI/blob/stash_simulate/src/coords.py#L61

Reply:
> ![gridding demo](https://user-images.githubusercontent.com/18370798/160550866-aaf3cbbd-699e-4fcf-970b-a4b458fda87a.PNG)


The last projection against (1,0,0) looks bunches at polls.

Reply:
![demo2](https://user-images.githubusercontent.com/18370798/160774746-9cd1c44e-d1a8-461d-b158-d78039a0f51e.PNG)

Fixed distributions using geomstats